Add rt1180 spi/edma support

lucien-nxp opened the pull request:
1. Add lpspi and edma instances on RT1180
2. Update edma driver to adapt EDMA4
3. Port lpspi_loopback case on RT1180, need wired connection shown below:
   /* To test this sample, connect J44.8 <-> J44.10 */
4. Fix uart case build issue, add overlay files for cm33/cm7 cores

lucien-nxp committed:
soc: nxp: imxrt: imxrt118x: add lpspi clock and trdc configuration

add lpspi clock enablement code

DMA3/4 access different domain is controlled by TRDC, release all
the domain access permission for DMA3/4, and add privilege and secure
information in dma access request signal by DAC module

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
dts: arm: nxp: rt118x: add lpspi and edma instances for RT118X

add lpspi and edma instances for RT118X
Find mpu description missed for RT1180X, add mpu in dts
add dmas controller setting for lpspi instances

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
driver: clock_control_mcux_ccm_rev2.c: update new lpspi support

On RT1180, two lpspis share the same clock root,
support lpspi in clock driver for RT118x

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
drivers: dma: dma_mcux_edma: support EDMA IP in edma drivers

Multi channels share one IRQ, add channels-shared-irq-mask on RT1180
attribution to describe the channel shared status, and add code
implementation to register the handler function for each channel
in different interrupts.

Fix legacy building warning issue

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
boards: nxp: mimxrt1180_evk: enable lpspi3 and edma4 instance support

Add pinctrl and dma channel/mux information for lpspi3 on cm33/cm7 cores.
Set dma4 status is OK.
Add pinctrl for lpuart3

tests: driver: spi: support spi_loopback case on MIMXRT1180_EVK
Add .config/.overlay files for cm33/cm7 cores
Link to ocram1 as RAM region on cm7 core
Link to dtcm as RAM region on cm33 core

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

PetervdPerk-NXP commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -651,10 +651,39 @@ static int dma_mcux_edma_init(const struct device *dev)
 			    irq_enable(DT_INST_IRQ_BY_IDX(n, idx, irq));	\
 	}
 
+#define DMA_CHANNLES_SHARED_IRQ_MASK(i, n) \
```
Typo

lucien-nxp replied:
Can you be a little more specific?

PetervdPerk-NXP replied:
I think `CHANNLES` should be `CHANNELS`

lucien-nxp replied:
Yes, thank you, I will correct it later.

dleach02 replied:
Does the build give an error on this?

lucien-nxp replied:
No, build error is caused by dma_mcux_edma_error_irq_handler function defined but not used. However, IRQ_CONNECT function will include the function into interrupt function. I don't know how to resolve the building warning issue, and it seems a legacy issue. (There is standalone error IRQ on RT1180 edma.)

bperseghetti reviewed (changes requested):
Currently edma is not properly supported with NXP's LPSPI in Zephyr since the hardware chip select is not currently working in the LPSPI driver.

lucien-nxp commented:
> Currently edma is not properly supported with NXP's LPSPI in Zephyr since the hardware chip select is not currently working in the LPSPI driver.

However, I enabled the lpspi_loopback case with edma and tested passed on my local. @dleach02 @danieldegrasse @decsny ,do you have any comment on lpspi_edma issue?
By the way, it seems the test building result is a problem left over by history.
  [Building Error](https://github.com/zephyrproject-rtos/zephyr/pull/77458/checks?check_run_id=29168337236)

bperseghetti commented:
> > Currently edma is not properly supported with NXP's LPSPI in Zephyr since the hardware chip select is not currently working in the LPSPI driver.
> 
> However, I enabled the lpspi_loopback case with edma and tested passed on my local. @dleach02 @danieldegrasse @decsny ,do you have any comment on lpspi_edma issue? By the way, it seems the test building result is a problem left over by history. [Building Error](https://github.com/zephyrproject-rtos/zephyr/pull/77458/checks?check_run_id=29168337236)

I think it would be good to consider using a test other than loopback for testing DMA with LPSPI that also tests HW CS.
I believe that this EVK has mikrobus and could use the in tree sensor streaming API with the ICM42688 using https://www.mikroe.com/6dof-imu-14-click?srsltid=AfmBOop3WDHNqW-m4sLp4pcR6l5OtbKUhniA-__iqOI2SMP6_6vGbf1D

Linking these here for tracking purposes:
- https://github.com/zephyrproject-rtos/zephyr/issues/16544
- https://github.com/zephyrproject-rtos/zephyr/issues/69658

danieldegrasse commented:
@lucien-nxp can you take a look at the CI failures? I think the kernel timer one might be fixed by rerunning the test (can't reproduce locally), but I believe the build failures in `tests/drivers/uart/uart_async_api` and `tests/drivers/spi/spi_loopback` need a second look

lucien-nxp commented:
> > > Currently edma is not properly supported with NXP's LPSPI in Zephyr since the hardware chip select is not currently working in the LPSPI driver.
> > 
> > 
> > However, I enabled the lpspi_loopback case with edma and tested passed on my local. @dleach02 @danieldegrasse @decsny ,do you have any comment on lpspi_edma issue? By the way, it seems the test building result is a problem left over by history. [Building Error](https://github.com/zephyrproject-rtos/zephyr/pull/77458/checks?check_run_id=29168337236)
> 
> I think it would be good to consider using a test other than loopback for testing DMA with LPSPI that also tests HW CS. I believe that this EVK has mikrobus and could use the in tree sensor streaming API with the ICM42688 using https://www.mikroe.com/6dof-imu-14-click?srsltid=AfmBOop3WDHNqW-m4sLp4pcR6l5OtbKUhniA-__iqOI2SMP6_6vGbf1D
> 
> Linking these here for tracking purposes:
> 
> * [drivers: spi: spi_mcux_lpspi: inconsistent chip select behaviour #16544](https://github.com/zephyrproject-rtos/zephyr/issues/16544)
> * [LAN865x not working with RT1170 & RT1064 (likely caused by LPSPI issues) #69658](https://github.com/zephyrproject-rtos/zephyr/issues/69658)

DMA case couldn't be supported on RT1180 because each dma request need to be triggered by software because there is no nxp,aon function like other RT platforms on RT1180. Another way to resolving the dma request is enabling channel link, however, limited to test case's configuration, it's hard to implemented on RT1180.

hakehuang commented:
board testing pass

decsny commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -667,6 +696,8 @@ static int dma_mcux_edma_init(const struct device *dev)
 	IRQ_CONFIG(n, idx, dma_mcux_edma_##n##_irq_##idx)
 
 #define DMA_MCUX_EDMA_CONFIG_FUNC(n)						\
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(n, channels_shared_irq_mask), \
```
this isn't technically what IF_ENABLED is for, you should probably use cond_code_1

lucien-nxp replied:
Why I should cond_code_1?  I don't need _else_code on this case. 
`#define COND_CODE_1(_flag, _if_1_code, _else_code) \
	Z_COND_CODE_1(_flag, _if_1_code, _else_code)`

decsny replied:
okay, on second thought, you can actually use this macro here since it is checking for a 1, nevermind

decsny commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -651,10 +651,39 @@ static int dma_mcux_edma_init(const struct device *dev)
+
+#define EDMA_CHANNELS_SHARED_REGISTER_IN_IRQ(idx, n, dev) \
+		uint32_t step = GET_EDMA_CHANNEL_SHARED_IRQ_STEP(n); \
+		for (int _i = 0; _i < step; _i++) {	\
```
can you make this macro define a function helper definition so that the preprocessor output is more readable when debugging , otherwise all this code will end up on one line with the rest of the init macro

lucien-nxp replied:
Can't use function to replace macro, because instance index n and  edma_channel_mask_##n array need to be used by EDMA_CHANNELS_SHARED_REGISTER_IN_IRQ. This is a common driver,  driver don't know how many edma instance will exist on one NPI.

lucien-nxp replied:
Updated.

decsny commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -214,7 +214,7 @@ static void dma_mcux_edma_irq_handler(const struct device *dev, uint32_t channel
 }
 
-#if !DT_INST_PROP(0, no_error_irq)
+#if !DT_INST_NODE_HAS_PROP(0, no_error_irq)
```
use DT_ANY_INST_HAS_PROP_STATUS_OKAY at least

lucien-nxp replied:
Updated.

decsny commented on dts/bindings/dma/nxp,mcux-edma-v4.yaml:
```diff
@@ -12,3 +12,16 @@ properties:
     type: boolean
     description: |
       If the SoCs don't have a separate interrupt id for error IRQ.
+  dma-channels:
```
who else is using this binding, is making this required true in this commit going to break bisectibility ?

decsny commented on boards/nxp/mimxrt1180_evk/mimxrt1180_evk_mimxrt1189_cm33.yaml:
```diff
@@ -19,4 +19,6 @@ supported:
   - i2c
   - counter
+  - dma
+  - spi
```
changing board yaml before commit that enables the testing overlays? is it intended ? (maybe it doesnt matter since it never was enabled before this)

lucien-nxp replied:
Squash the board and test modification related to spi_edma commit into one commit.

decsny commented on boards/nxp/mimxrt1180_evk/doc/index.rst:
need to update the connections/io's section

lucien-nxp replied:
Updated.

EmilioCBen reviewed (commented):
Can you update the last commit message to instead reference the overlays being added for the uart_async_api test rather then claiming an issue will be assigned?
Other than that I think this PR is good.

lucien-nxp commented:
> Can you update the last commit message to instead reference the overlays being added for the uart_async_api test rather then claiming an issue will be assigned? Other than that I think this PR is good.

Thank you. Updated.

dleach02 commented:
resolve the merge conflict

lucien-nxp commented on soc/nxp/imxrt/imxrt118x/soc.c:
```diff
@@ -293,13 +313,6 @@ static ALWAYS_INLINE void clock_init(void)
-	/* Enable the AHB clock while the CM7 is sleeping to allow debug access
-	 * to TCM
-	 */
-	BLK_CTRL_S_AONMIX->M7_CFG |= BLK_CTRL_S_AONMIX_M7_CFG_TCM_SIZE_MASK;
```
This code shouldn't have been added in the first place. We can't modify the default value recorded in M7_CFG, otherwise, edma driver can't read correct data from this register and do correct address transition.

danieldegrasse commented:
@lucien-nxp from what I can tell the issue here is caused by not providing a `zephyr,dtcm` chosen node for the CM7 build target, like we do for other CM7 cores: https://github.com/zephyrproject-rtos/zephyr/blob/2f800cea8fdf48f8e58add5b9ebd12aed96a6113/boards/nxp/mimxrt1064_evk/mimxrt1064_evk.dts#L33

lucien-nxp committed:
tests: drivers: uart: uart_async_api: fix build error for RT1180

Due to cm7 on RT1180 is secondary core, and dtcm have been used
as sram directly, so there is no need to define zephyr,dtcm node.

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

danieldegrasse commented on tests/drivers/spi/spi_loopback/boards/mimxrt1180_evk_mimxrt1189_cm7.conf:
```diff
@@ -0,0 +1,8 @@
+#
+CONFIG_SPI_MCUX_LPSPI_DMA=y
+CONFIG_SPI_ASYNC=n
+CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS=n
```
Why turn this off for the CM7? We have DTCM available, right?

lucien-nxp replied:
Yes, DTCM is available. However, dtcm region have been set to used as zephyr,sram node on cm7 core directly.

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -275,6 +274,27 @@ static void dma_mcux_edma_error_irq_handler(const struct device *dev)
+			num = num >> 1;
+			count++;
+		}
+		index += _i;
```
Why add to index like this? Wouldn't we simply want to increment index by 1 each iteration? We could also remove index entirely and simply move the `buf` pointer.

lucien-nxp replied:
Updated. Could you help check again?

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -275,6 +274,27 @@ static void dma_mcux_edma_error_irq_handler(const struct device *dev)
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(channels_shared_irq_mask)
+static void dma_mcux_edma_multi_channels_irq_handler(const struct device *dev, uint32_t idx,
+			uint32_t *buf, uint32_t step)
```
I'd suggest renaming the `step` variable to `mask_width` or something similar- that might make this code block a bit easier to follow IMO.

lucien-nxp replied:
Update to use mask_width variable and modify the related macro name.

danieldegrasse commented on dts/bindings/dma/nxp,mcux-edma.yaml:
```diff
@@ -65,6 +65,18 @@ properties:
+      The channel number is mapped to the bit value of array element value.
+      If the interrupt is shared on one channel number, the correspongding
+      bit is set to 1.
+      Please note each element of the array must be 32-bit. If there are more
```
Might be worth specifying that the elements should be contiguous for the case where more than 32 channels exist

lucien-nxp replied:
Updated.

lucien-nxp commented on drivers/dma/dma_mcux_edma.c:
Moved.

lucien-nxp commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -856,10 +876,28 @@ static int dma_mcux_edma_init(const struct device *dev)
+#define EDMA_CHANNELS_SHARED_IRQ_MASK(i, n) \
+		(DT_PROP_BY_IDX(DT_DRV_INST(n), channels_shared_irq_mask, i))
+
+#define EDMA_CHANNELS_MASK(n) static uint32_t edma_channel_mask_##n[] = { \
```
Yes, I have simplified it to get array.

lucien-nxp commented:
Only rebase to the newest commit and resolve the conflicts.

lucien-nxp commented:
Only rebase to the newest commit and resolve the conflicts.

lucien-nxp commented:
Rebase to [#82761](https://github.com/zephyrproject-rtos/zephyr/pull/82761), resolve the tpm building issues.

hakehuang commented:
@PerMac  the bsim CI failure is from latest main line code, see
https://github.com/zephyrproject-rtos/zephyr/issues/82835

lucien-nxp commented:
Only rebase the main branch newest commit and resolve the conflicts.

lucien-nxp commented:
Anyone can help to merge the PR?

decsny commented:
need assignee approval

lucien-nxp commented:
@tbursztyka , could you help approve and merge the PR?
